selfdrive/ui/replay/filereader.cc

```diff
@@ -1,139 +1,132 @@
-  bStream.next_in = NULL;
-  bStream.avail_in = 0;
-  bStream.bzalloc = NULL;
-  bStream.bzfree = NULL;
-  bStream.opaque = NULL;
-
-  int ret = BZ2_bzDecompressInit(&bStream, 0, 0);
-  if (ret != BZ_OK) qWarning() << "bz2 init failed";
+// class LogReader
 
-  // start with 64MB buffer
-  raw.resize(1024*1024*64);
-
-  // auto increment?
-  bStream.next_out = raw.data();
-  bStream.avail_out = raw.size();
-
-  // parsed no events yet
-  event_offset = 0;
-
-  parser = new std::thread([&]() {
-    while (1) {
-      mergeEvents(cdled.get());
-    }
+LogReader::LogReader(const QString &file, QObject *parent) : QObject(parent) {
+  file_reader_ = new FileReader(file);
+  file_reader_->moveToThread(&thread_);
+  connect(&thread_, &QThread::started, file_reader_, &FileReader::read);
+  connect(&thread_, &QThread::finished, file_reader_, &FileReader::deleteLater);
+  connect(file_reader_, &FileReader::finished, [=](const QByteArray &dat) {
+    parseEvents(dat);
   });
+  connect(file_reader_, &FileReader::failed, [=](const QString &err) {
+    qDebug() << err;
+  });
+  thread_.start();
 }
 
 LogReader::~LogReader() {
-  delete parser;
+  // wait until thread is finished.
+  exit_ = true;
+  file_reader_->abort();
+  thread_.quit();
+  thread_.wait();
+
+  // clear events
+  for (auto e : events) {
+    delete e;
+  }
 }
 
-void LogReader::mergeEvents(int dled) {
-  auto amsg = kj::arrayPtr((const capnp::word*)(raw.data() + event_offset), (dled-event_offset)/sizeof(capnp::word));
-  Events events_local;
-  QMap<int, QPair<int, int> > eidx_local;
-
-  while (amsg.size() > 0) {
-    try {
-      capnp::FlatArrayMessageReader cmsg = capnp::FlatArrayMessageReader(amsg);
-
-      // this needed? it is
-      capnp::FlatArrayMessageReader *tmsg =
-        new capnp::FlatArrayMessageReader(kj::arrayPtr(amsg.begin(), cmsg.getEnd()));
-
-      amsg = kj::arrayPtr(cmsg.getEnd(), amsg.end());
+void LogReader::parseEvents(const QByteArray &dat) {
+  raw_.resize(1024 * 1024 * 64);
+  if (!decompressBZ2(raw_, dat.data(), dat.size())) {
+    qWarning() << "bz2 decompress failed";
+  }
 
-      cereal::Event::Reader event = tmsg->getRoot<cereal::Event>();
-      events_local.insert(event.getLogMonoTime(), event);
+  auto insertEidx = [&](CameraType type, const cereal::EncodeIndex::Reader &e) {
+    eidx[type][e.getFrameId()] = {e.getSegmentNum(), e.getSegmentId()};
+  };
 
-      // hack
-      // TODO: rewrite with callback
-      if (event.which() == cereal::Event::ROAD_ENCODE_IDX) {
-        auto ee = event.getRoadEncodeIdx();
-        eidx_local.insert(ee.getFrameId(), qMakePair(ee.getSegmentNum(), ee.getSegmentId()));
+  valid_ = true;
+  kj::ArrayPtr<const capnp::word> words((const capnp::word *)raw_.data(), raw_.size() / sizeof(capnp::word));
+  while (!exit_ && words.size() > 0) {
+    try {
+      std::unique_ptr<Event> evt = std::make_unique<Event>(words);
+      switch (evt->which) {
```

Comment:
why not clean this up with capnp's dynamic api?

Reply:
here's an example: https://github.com/commaai/openpilot/blob/master/selfdrive/camerad/cameras/camera_frame_stream.cc#L47

Reply:
Because I think it's faster and straightforward